Blizzard.IO.RabbitMQ/NetqRabbitConsumer.cs

```diff
@@ -0,0 +1,93 @@
+    {
+        private readonly IBus _netqBus;
+        private readonly IDeserializer<TData> _deserializer;
+        private readonly IConcreteTypeDeserializer<TData> _concreteTypeDeserializer;
+        private readonly IQueue _sourceQueue;
+        private readonly IConverter<MessageProperties, RabbitMessageProperties> _coverter;
+        private readonly ILogger<NetqRabbitConsumer<TData>> _logger;
+        
+        private IDisposable _consumeTask;
+        private readonly Action<byte[], MessageProperties, MessageReceivedInfo> _consumingStrategy;
+
+        public event Action<TData> MessageReceived;
+        public event Action<TData, RabbitMessageProperties> MessageWithMetadataReceived;
+
+        protected NetqRabbitConsumer(IBus netqBus, IQueue sourceQueue, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> coverter)
+        {
+            _netqBus = netqBus;
+            _sourceQueue = sourceQueue;
+            _logger = loggerFactory.CreateLogger<NetqRabbitConsumer<TData>>();
+            _coverter = coverter ?? new RabbitPropertiesConverter();
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, IQueue sourceQueue, IDeserializer<TData> deserializer, 
+            ILoggerFactory loggerFactory, IConverter<MessageProperties, RabbitMessageProperties> coverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, coverter)
+        {
+            _deserializer = deserializer;
+            _consumingStrategy += ConsumeViaDeserializer;
+            _logger.LogDebug($"Initialized consumer from queue: {_sourceQueue.Name}," +
+                $" expecting {typeof(TData).ToString()} data");
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, IQueue sourceQueue, IConcreteTypeDeserializer<TData> concreteTypeDeserializer,
+            ILoggerFactory loggerFactory, IConverter<MessageProperties, RabbitMessageProperties> coverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, coverter)
+        {
+            _concreteTypeDeserializer = concreteTypeDeserializer;
+            _consumingStrategy += ConsumeViaConcreteDeserializer;
+            _logger.LogDebug($"Initialized consumer from queue: {_sourceQueue.Name}," +
+                $" expecting {typeof(TData).ToString()} derieved data");
+        }
+
+        public void Start()
+        {
+            _consumeTask = _netqBus.Advanced.Consume(_sourceQueue, _consumingStrategy);
+            _logger.LogInformation("Started consumer sucessfully");
+        }
+
+        private void ConsumeViaDeserializer(byte[] messageBytes, MessageProperties messageProperties,
+            MessageReceivedInfo messageInfo)
+        {
+            TData data = _deserializer.Deserialize(messageBytes);
+            RabbitMessageProperties properties = _coverter.Convert(messageProperties);
+            MessageReceived?.Invoke(data);
+            MessageWithMetadataReceived?.Invoke(data, properties);
```

Comment:
you still have code duplication on those three lines.
I suggest to extract those lines to a lambda in the start fucntion that will call the strategy function and then execute those lines.

Reply:
f